include/NovelRT/DotNet/RuntimeService.h

```diff
@@ -8,41 +8,36 @@
 #endif
 
 namespace NovelRT::DotNet {
-  class RuntimeService {
+  class RuntimeService : public std::enable_shared_from_this<RuntimeService> {
   private:
+    struct Exports {
+      void(*Initialise)();
+      void(*Teardown)();
+      void(*FreeObject)(intptr_t obj);
+      void(*FreeString)(const char* str);
+      void(*GetInkServiceExports)(struct Ink::InkService::Exports* exports);
+    };
+
     Utilities::Lazy<hostfxr_handle> _hostContextHandle;
     Utilities::Lazy<void*> _hostfxr;
     Utilities::Lazy<hostfxr_initialize_for_runtime_config_fn> _hostfxr_initialize_for_runtime_config;
     Utilities::Lazy<hostfxr_get_runtime_delegate_fn> _hostfxr_get_runtime_delegate;
     Utilities::Lazy<hostfxr_close_fn> _hostfxr_close;
     Utilities::Lazy<load_assembly_and_get_function_pointer_fn> _load_assembly_and_get_function_pointer;
+    Utilities::Lazy<Exports> _exports;
     LoggingService _logger;
 
   public:
     RuntimeService();
     ~RuntimeService();
 
-    template <class _Fty>
-    std::function<_Fty> getFunction(const char_t* assemblyName, const char_t* typeName, const char_t* methodName, const char_t* delegateTypeName) {
-      std::filesystem::path executableDirPath = Utilities::Misc::getExecutableDirPath();
-      std::filesystem::path assemblyPath = executableDirPath / "dotnet" / assemblyName;
-
-      const char_t* assembly_path = assemblyPath .c_str();
-
-      void* delegate;
-      int result = _load_assembly_and_get_function_pointer.getActual()(assembly_path, typeName, methodName, delegateTypeName, nullptr, &delegate);
-
-      if (result != 0)
-      {
-        _logger.logError("Failed to locate the specified managed function: {}", result);
-        throw std::runtime_error("Failed to locate the specified managed function");
-      }
-
-      return std::function<_Fty>(reinterpret_cast<_Fty*>(delegate));
-    }
-
```

Comment:
What happened here? I thought the ink stuff was just gonna sit on top of the runtime servie in our plan to standardise the runtime service as an interface that other runtimes such as JVM can have bindings written for? Was this method just redesigned?

Reply:
This was exposing the 3.1 `getFunction` which required matching delegate types. The new functionality exposes the 5.0 `getFunction` which just returns a straight function pointer to a `UnmanagedCallersOnly` static method.

Reply:
Ahhh so it's done on the C# side effectively?

Reply:
Not quite. Its that function pointers didn't exist in .NET Core 3.1 but they do in .NET 5 so the API we need in NovelRT changed.
This reflects that by using the newer/better/cheaper function pointer API instead and removing the older one that required a delegate and extra allocation.

Reply:
I see. Thanks!